### ISSUE TITLE
Abysmal performance in narrowed buffer

### ISSUE THREAD

**Issue by Fuco1** [emacs-lsp/lsp-mode#977]
**Describe the bug**

Unusable performance in a narrowed buffer, Emacs freezes and becomes unresponsive for minutes+

**To Reproduce**

1. Get a 1000+ line php buffer
2. Narrow down to a function
3. Insert any text

**Expected behavior**

There should be no noticable difference in a narrowed vs unnarrowed buffer.

I suspect you might be using the "standard idiom"

```elisp
(save-excursion
  (save-restriction
    (widen)
    ;; your code goes here
    ))
```

in some super-tight loop.  This has killed me several times before in both smartparens and Elsa.  Both `save-restriction` and `widen` kill performance if used in a loop.

**Which Language Server did you use**

https://github.com/felixfbecker/php-language-server latest commit

**OS**

```
Linux thales 4.15.0-20-generic #21-Ubuntu SMP Tue Apr 24 06:16:15 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by sebastiansturm** [comment]
I may be seeing something related now (though I believe that didn't use to be the case until recently): in a C++ file of about 2000 lines, `evil-insert-newline-below` takes ridiculously long to execute (~0.5 s per invocation). Doesn't seem to matter where I am within the buffer. I manually instrumented some lsp methods containing the save-widen-idiom mentioned by @Fuco1, but found no smoking gun. @Fuco1: do you have any further hints on how to find the root case? Instrumenting all of `lsp` using `elp-instrument-package` didn't turn up anything particularly slow either

**Comment by sebastiansturm** [comment]
also, whatever I'm seeing is not deterministic. Restarting Emacs with identical settings has fixed the performance issue for now (I'm sure it will come back though). I noticed, however, that the buffer now contains 955 overlays as opposed to ~6500 just before restart. So maybe my experimental semantic-highlighting handler leaks overlays, will have to add some instrumentation I guess

**Comment by yyoncho** [comment]
@sebastiansturm does it happen in a narrowed buffer? You may try also `profile-start`/`profile-report`. As a side note,  `gc-cons-threshold` not being high enough may cause such problems and it wont show if you have introspected only `lsp-mode` package.

**Comment by sebastiansturm** [comment]
hm, some of what I'm seeing may be related to narrowing, most of it probably isn't. I didn't originally use narrowing in that buffer and I just noticed that narrowing actually improves performance in my case (so I guess I'll better open a separate issue for the overlay thing). OTOH, widening afterwards then caused Emacs to (seemingly) freeze. Got to go now, so I can't investigate this in detail right now but interrupting Emacs via SIGUSR2 tells me it's currently within `symbol-overlay-refresh`.
My `gc-cons-threshold` is rather huge (100 MB); will check the profiler again tomorrow or so though last time I used it the main hotspot was one of the `evil-open-...` functions (was a leaf node of the call graph, so I couldn't zoom in further)
